### PR TITLE
Implement CA1816 Call GC.SuppressFinalize correctly rule.

### PR DESCRIPTION
**Pull request by balajikris**
This implementation is fully backwards compatible with existing FxCop implementation of the rule.
Plus, unit tests for C# and VB.

### CODE REVIEW

File: src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs [hunk 1]
```diff
@@ -27,42 +29,239 @@ public abstract class CallGCSuppressFinalizeCorrectlyAnalyzer : DiagnosticAnalyz
...
+
+                    operationBlockContext.RegisterOperationAction(analyzer.Analyze, OperationKind.InvocationExpression);
+
+                    operationBlockContext.RegisterOperationBlockEndAction(analyzer.OperationBlockEndAction);
+                });
+            });
+
+        }
+
+
```
Comment (balajikris) [inline src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs]: will remove additional blank lines

**Comment by balajikris** [comment]
tagging @dotnet/roslyn-analysis for review.

File: src/System.Runtime.Analyzers/UnitTests/CallGCSuppressFinalizeCorrectlyTests.cs [hunk 2]
```diff
@@ -16,5 +30,568 @@ protected override DiagnosticAnalyzer GetCSharpDiagnosticAnalyzer()
         {
             return new CSharpCallGCSuppressFinalizeCorrectlyAnalyzer();
```
Comment (mavasani) [inline src/System.Runtime.Analyzers/UnitTests/CallGCSuppressFinalizeCorrectlyTests.cs]: Did you add any code to the C# and VB specific derived types for this analyzer? Otherwise, please delete them, change CallGCSuppressFinalizeCorrectlyAnalyzer to be a sealed type with DiagnsoticAnalyzerAttribute(C#, VB).

File: src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs [hunk 3]
```diff
@@ -27,42 +29,239 @@ public abstract class CallGCSuppressFinalizeCorrectlyAnalyzer : DiagnosticAnalyz
                                                                              s_localizableMessageNotCalledWithFinalizer,
                                                                              DiagnosticCategory.Performance,
```
Comment (mavasani) [inline src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs]: This is an FXCop Usage rule. @srivatsn @nguerrera did we decide to make it a performance rule?

File: src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs [hunk 4]
```diff
@@ -27,42 +29,239 @@ public abstract class CallGCSuppressFinalizeCorrectlyAnalyzer : DiagnosticAnalyz
...
+            public SuppressFinalizeAnalyzer(IMethodSymbol methodSymbol, Compilation compilation)
+            {
+                this._compilation = compilation;
+                this._containingMethodSymbol = methodSymbol;
+            }
+
+            public void Analyze(OperationAnalysisContext analysisContext)
+            {
+                _invocationExpression = (IInvocationExpression)analysisContext.Operation;
+                _gcSuppressFinalizeMethodSymbol = _compilation.GetTypeByMetadataName("System.GC").GetMembers("SuppressFinalize").OfType<IMethodSymbol>().SingleOrDefault();
```
Comment (mavasani) [inline src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs]: You should compute well known types/members in the CompilationStartAction, avoid registering any operation block action if gcSuppressFinalizeMethodSymbol == null, and if not store it as a readonly field of SuppressFinalizeAnalyzer.

File: src/System.Runtime.Analyzers/UnitTests/CallGCSuppressFinalizeCorrectlyTests.cs (hunk 2 above)
Comment (balajikris) [inline src/System.Runtime.Analyzers/UnitTests/CallGCSuppressFinalizeCorrectlyTests.cs]: No language specific code here, will clean up the stubs.

File: src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs (hunk 4 above)
Comment (balajikris) [inline src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs]: thanks, good suggestion.

File: src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs [hunk 5]
```diff
@@ -27,42 +30,221 @@ public abstract class CallGCSuppressFinalizeCorrectlyAnalyzer : DiagnosticAnalyz
...
+            public SuppressFinalizeAnalyzer(IMethodSymbol methodSymbol, IMethodSymbol gcSuppressFinalizeMethodSymbol, Compilation compilation)
+            {
+                this._compilation = compilation;
+                this._containingMethodSymbol = methodSymbol;
+                this._gcSuppressFinalizeMethodSymbol = gcSuppressFinalizeMethodSymbol;
+            }
+
+            public void Analyze(OperationAnalysisContext analysisContext)
+            {
+                _invocationExpression = (IInvocationExpression)analysisContext.Operation;
```
Comment (mavasani) [inline src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs]: You don't need to store the invocationExpression onto a field. I would recommend following change:

1. Make invocationExpression a local.
2. Move the entire block from line 156-181 inside the if block at the end of this method, i.e. line 134 - they should be diagnostics reported on the invocation Operation, not in OperationBlockEndAction.
3. Can you compute and store the semantic model only for the first time, i.e. if it is null?

Comment (mavasani) [inline src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs]: You can invoke GetAllowedSuppressFinalizeUsage in the constructor of SuppressFinalizeAnalyzer and store the result onto a readonly field.

File: src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs [hunk 6]
```diff
@@ -27,42 +30,221 @@ public abstract class CallGCSuppressFinalizeCorrectlyAnalyzer : DiagnosticAnalyz
...
+                // Even if the Dispose method is declared on a type without a
+                // finalizer, we still require it to call GC.SuppressFinalize to
+                // prevent derived finalizable types from having to reimplement
+                // IDisposable.Dispose just to call it.
+                return SuppressFinalizeUsage.MustCall;
+            }
+
+            private bool IsDisposeMethod(IMethodSymbol method)
+            {
+                var disposableType = WellKnownTypes.IDisposable(_compilation);
```
Comment (mavasani) [inline src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs]: Compute disposableType in the CompilationStartAction and store it into a readonly field here.

File: src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs [hunk 7]
```diff
@@ -27,42 +30,221 @@ public abstract class CallGCSuppressFinalizeCorrectlyAnalyzer : DiagnosticAnalyz
...
+                }
+
+                // Even if the Dispose method is declared on a type without a
+                // finalizer, we still require it to call GC.SuppressFinalize to
+                // prevent derived finalizable types from having to reimplement
+                // IDisposable.Dispose just to call it.
+                return SuppressFinalizeUsage.MustCall;
+            }
+
+            private bool IsDisposeMethod(IMethodSymbol method)
```
Comment (mavasani) [inline src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs]: Seems like a good candidate to move into AnalyzerUtilities.IMethodSymbolExtensions?

File: src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs (hunk 5 above)
Comment (balajikris) [inline src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs]: ah, i have a vague feeling that the API has changed since I initially wrote this, I remember doing an OperationBlockEnd just to be able to report diagnostics. Will make the change.

**Comment by balajikris** [comment]
VB Tests are in now and all PR feedback have been addressed. The analyzer implementation and tests are now complete, while the fixer is still NYI. Let me know what the merge policies are (how many sign off required) and I will merge after they are satisfied.

File: src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs [hunk 8]
```diff
@@ -27,42 +30,192 @@ public abstract class CallGCSuppressFinalizeCorrectlyAnalyzer : DiagnosticAnalyz
...
+                    {
+                        return;
+                    }
+
+                    var parameterSymbol = _semanticModel.GetSymbolInfo(invocationExpression.ArgumentsInSourceOrder.Single().Syntax).Symbol as IParameterSymbol;
+                    if (parameterSymbol == null || !parameterSymbol.IsThis)
+                    {
+                        analysisContext.ReportDiagnostic(invocationExpression.Syntax.CreateDiagnostic(
+                            NotPassedThisRule,
+                            _containingMethodSymbol.ToDisplayString(SymbolDisplayFormats.ShortSymbolDisplayFormat),
```
Comment (mavasani) [inline src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs]: We generally avoid displaying full symbol names in the diagnostic messages - we should just use symbol.Name here.

**Comment by mavasani** [comment]
:+1: modulo changing the diagnostic messages to use symbol.Name.

File: src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs (hunk 8 above)
Comment (balajikris) [inline src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs]: This is how existing FxCop library is displaying the diagnostic, So, I kept the behavior. I personally find it useful to qualify with the type name here but let me know if you feel strongly the other way and I'll change.

Comment (mavasani) [inline src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs]: A while back @heejaechang cleaned up compiler diagnostics to avoid fully qualified names in diagnostic messages, and we try to maintain that approach for analyzer diagnostics. It would be good if you can do it, otherwise we can do it as a part of a cleanup pass when we scan all analyzers - anything is fine.

Comment (balajikris) [inline src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs]: if you're eventually going to change it anyway, I guess I can just do it now.

File: src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs (hunk 3 above)
Comment (srivatsn) [inline src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs]: Nope, the stub generator seems to have gotten it wrong. Please change it. Also can edit the System.Runtime.Analyzers.props file in this project and add a -Microsoft.Usage#CA1816 entry to it. That turns off the FxCop implementation when this rule runs.

File: src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs (hunk 8 above)
Comment (balajikris) [inline src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs]: spoke offline with @mavasani about this, using `symbol.Name` will not be sufficient as it will be `.ctor` for constructors and so on. We likely need to tweak `SymbolDisplayFormats.ShortSymbolDisplayFormat` to omit type name qualification but that will require cleaning up all test baselines. I'm going to leave this as such in favor of a bulk clean up pass later.

Comment (srivatsn) [inline src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs]: Most often people view the error list when they have context and so short names tend to be enough to disambiguate. Fully qualified names tend to make the message much longer and make the error unintelligible.

File: src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs (hunk 3 above)
Comment (balajikris) [inline src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs]: will do. I merged after speaking with manish offline, will have to send another PR :\
